src/runtime/runtime.cpp

```diff
@@ -167,6 +167,24 @@ namespace
       context->notify(error.c_str(), context->data, 0, NULL);
     }
   }
+
+  void releaseCommand(oclgrind::Command *command)
```

Comment:
I don't quite understand why this is needed (along with the `execBefore` list). Why can't we just release command as soon as it is executed? Dependent commands will have a reference to the event, which will still be accessible after its corresponding command has been released.

Reply:
I did this: a) to preserve the existing logic of creating and releasing command objects in the runtime; and b) to separate the (recursive) execution of a command and its dependencies from the (recursive) release of said command and those same dependencies. We can easily release commands in the `Queue` class after execution, but it doesn't seem so clean (although probably simpler in terms of code). Should I do this, or leave as it is?

Reply:
Happy to leave this as is for now - we can revisit at a later date if needed.